oshi-core-java25/src/main/java/oshi/ffm/driver/windows/registry/InstalledAppsDataFFM.java

```diff
@@ -0,0 +1,221 @@
+                                LOG.trace("Skipping key {}: {}", fullPath, e.getMessage());
+                            }
+                        }
+                    } catch (Throwable e) {
+                        LOG.trace("Skipping path {}: {}", registryPath, e.getMessage());
+                    }
+                }
+            }
+        }
+
+        return new ArrayList<>(appInfoSet);
+    }
+
+    private static String registryValueToString(Object registryValueOrNull) {
+        if (registryValueOrNull instanceof Integer i) {
+            return Integer.toString(i);
+        }
+        return (String) registryValueOrNull;
+    }
+
+    private static long registryValueToLong(Object registryValueOrNull) {
+        if (registryValueOrNull == null) {
+            return 0L;
+        }
+        long currentTimeSecs = System.currentTimeMillis() / 1000L;
+        long minSaneTimestamp = currentTimeSecs - THIRTY_YEARS_IN_SECS;
+
+        if (registryValueOrNull instanceof Integer i) {
+            if (i > minSaneTimestamp && i < currentTimeSecs) {
+                return i * 1000L;
+            }
+            return i;
+        } else if (registryValueOrNull instanceof String s) {
+            String dateStr = s.trim();
+            long epoch = ParseUtil.parseDateToEpoch(dateStr, "yyyyMMdd");
+            if (epoch == 0) {
+                epoch = ParseUtil.parseDateToEpoch(dateStr, "MM/dd/yyyy");
+            }
+            return epoch;
+        }
+        return 0L;
+    }
+
+    private static Object getRegistryValueOrNull(MemorySegment rootKey, String path, String key, int accessFlag)
+            throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment phkKey = arena.allocate(ADDRESS);
+            int rc = RegOpenKeyEx(rootKey, toWideString(arena, path), 0, KEY_READ | accessFlag, phkKey);
+            if (rc != ERROR_SUCCESS) {
+                throw new Win32Exception(rc);
+            }
+
+            MemorySegment hKey = phkKey.get(ADDRESS, 0);
+            try {
+                return registryGetValue(hKey, key);
+            } finally {
+                rc = RegCloseKey(hKey);
+                if (rc != ERROR_SUCCESS) {
+                    throw new Win32Exception(rc);
+                }
+            }
+        } catch (Win32Exception e) {
+            LOG.trace("Unable to access " + path + " with flag " + accessFlag + ": " + e.getMessage());
+            return null;
+        }
+    }
+
+    private static Object registryGetValue(MemorySegment hKey, String valueName) throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment lpType = arena.allocate(JAVA_INT);
+            MemorySegment lpcbData = arena.allocate(JAVA_INT);
+
+            int rc = RegQueryValueEx(hKey, toWideString(arena, valueName), 0, lpType, MemorySegment.NULL, lpcbData);
+            if (rc != ERROR_SUCCESS && rc != ERROR_INSUFFICIENT_BUFFER) {
```

Comment:
This complicates my suggestion of a `checkSuccess` method, and now I remember why I didn't do it earlier.  So maybe leave it as is for now.  But an idea for later.